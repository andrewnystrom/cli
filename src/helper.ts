import chalk from 'chalk'
import { capitalize, forEach, isEmpty, omit, padEnd } from 'lodash'

import {
  CommandsTreeNode,
  ICommandFunction,
  ICommandsTree
} from './interpreter'
import { ILogger } from './logger'
import { optionToString } from './utils'

export interface IOptionsAnnotations {
  [optionName: string]: string
}

export interface IHelpDetailedAnnoations {
  description?: string
  params?: string[]
  options?: IOptionsAnnotations
  [key: string]: any
}

export interface IPrintHelpArguments {
  node: CommandsTreeNode
  namespace: string
  logger: ILogger
}

interface IPrintCommandHelpArguments {
  command: ICommandFunction
  namespace: string
  logger: ILogger
}

interface IPrintNamespaceHelpArguments {
  commands: ICommandsTree
  namespace: string
  logger: ILogger
}

export type HelpAnnotations = string | IHelpDetailedAnnoations

export type PrintHelp = (args: IPrintHelpArguments) => void

function printCommandHelp({
  command,
  namespace,
  logger
}: IPrintCommandHelpArguments) {
  if (!command.help) {
    logger.log('Documentation not found')
    return
  }

  let annotations: IHelpDetailedAnnoations = {}

  if (typeof command.help === 'string') {
    annotations.description = command.help
  } else {
    annotations = command.help
  }

  const { description, params, options } = annotations
  const extra = omit(annotations, ['description', 'params', 'options'])
  const usageOptions = isEmpty(options) ? '' : '[options]'
  const usageParams =
    !Array.isArray(params) || isEmpty(params) ? '' : `[${params.join(' ')}]`

  logger.log(`Usage: ${namespace} ${usageOptions} ${usageParams}\n`)

  if (description) {
    logger.log(`${description}\n`)
  }

  if (!isEmpty(options)) {
    logger.log('Options:\n')
    forEach(options, (value, key) => {
      logger.log(`  ${padEnd(optionToString(key), 12)}${value}`)
    })
  }

  forEach(extra, (value, key) => {
    logger.log(`\n${capitalize(key)}:\n`)
    logger.log(`${value}\n`)
  })
}

function printNamespaceHelp({
  commands,
  namespace,
  logger
}: IPrintNamespaceHelpArguments) {
  Object.keys(commands)
    .sort()
    .forEach(key => {
      const node = commands[key]
      const nextNamespace = namespace ? `${namespace}:${key}` : key

      if (typeof node === 'function') {
        let annotations = node.help

        if (typeof annotations === 'string') {
          annotations = { description: annotations }
        }
        // Add task name
        const funcParams = annotations && annotations.params
        const logArgs = [chalk.bold(nextNamespace)]

        // Add task params
        if (Array.isArray(funcParams) && funcParams.length) {
          logArgs[0] += ` [${funcParams.join(' ')}]`
        }

        // Add description
        if (annotations && annotations.description) {
          const description = annotations.description
          logArgs[0] = padEnd(logArgs[0], 40) // format
          logArgs.push('-', description.split('\n')[0])
        }

        // Log
        logger.log(...logArgs)
      } else if (typeof node === 'object') {
        printNamespaceHelp({ commands: node, logger, namespace: nextNamespace })
      }
    })
}

export function printHelp({ node, namespace, logger }: IPrintHelpArguments) {
  if (typeof node === 'function') {
    printCommandHelp({ command: node, namespace, logger })
  } else {
    printNamespaceHelp({ commands: node, namespace, logger })
  }
}

export function help(command: ICommandFunction, annotations: HelpAnnotations) {
  // Because the validation above currently gets compiled out,
  // Explictly  validate the function input
  if (typeof command === 'function') {
    command.help = annotations
  } else {
    throw new Error('First help() argument must be a function')
  }
}
