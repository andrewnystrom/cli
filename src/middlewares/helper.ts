import chalk from 'chalk'
import { capitalize, forEach, isEmpty, omit, padEnd } from 'lodash'

import { CommandFunction, ICommandsDictionary, Middleware } from '../index'
import { ILogger } from '../utils/logger'
import { optionToString } from '../utils/options'

export interface IOptionsAnnotations {
  [optionName: string]: string
}

export interface IHelpDetailedAnnoations {
  description?: string
  params?: string[]
  options?: IOptionsAnnotations
  [key: string]: any
}

export type HelpAnnotations = string | IHelpDetailedAnnoations

export const annotationsMap = new Map<CommandFunction, HelpAnnotations>()

function printCommandHelp(
  command: CommandFunction,
  namespace: string,
  logger: ILogger
) {
  const help = annotationsMap.get(command)
  if (!help) {
    logger.log('Documentation not found')
    return
  }

  let annotations: IHelpDetailedAnnoations = {}

  if (typeof help === 'string') {
    annotations.description = help
  } else {
    annotations = help
  }

  const { description, params, options } = annotations
  const extra = omit(annotations, ['description', 'params', 'options'])
  const usageOptions = isEmpty(options) ? '' : '[options]'
  const usageParams =
    !Array.isArray(params) || isEmpty(params) ? '' : `[${params.join(' ')}]`

  logger.log(`Usage: ${namespace} ${usageOptions} ${usageParams}\n`)

  if (description) {
    logger.log(`${description}\n`)
  }

  if (!isEmpty(options)) {
    logger.log('Options:\n')
    forEach(options, (value, key) => {
      logger.log(`  ${padEnd(optionToString(key), 12)}${value}`)
    })
  }

  forEach(extra, (value, key) => {
    logger.log(`\n${capitalize(key)}:\n`)
    logger.log(`${value}\n`)
  })
}

function printDefinitionHelp(
  definition: ICommandsDictionary,
  namespace: string,
  logger: ILogger
) {
  if (!namespace) {
    logger.log('\nCommands:\n')
  }
  Object.keys(definition)
    .sort()
    .forEach(key => {
      const node = definition[key]
      const nextNamespace = namespace ? `${namespace}:${key}` : key

      if (typeof node === 'function') {
        let annotations = annotationsMap.get(node)

        if (typeof annotations === 'string') {
          annotations = { description: annotations }
        }
        // Add task name
        const funcParams = annotations && annotations.params
        const logArgs = [chalk.bold(nextNamespace)]

        // Add task params
        if (Array.isArray(funcParams) && funcParams.length) {
          logArgs[0] += ` [${funcParams.join(' ')}]`
        }

        // Add description
        if (annotations && annotations.description) {
          const description = annotations.description
          logArgs[0] = padEnd(logArgs[0], 40) // format
          logArgs.push('-', description.split('\n')[0])
        }

        // Log
        logger.log(...logArgs)
      } else if (typeof node === 'object') {
        printDefinitionHelp(node, nextNamespace, logger)
      }
    })
}

export function help(command: CommandFunction, annotations: HelpAnnotations) {
  // Because the validation above currently gets compiled out,
  // Explictly  validate the function input
  if (typeof command === 'function') {
    annotationsMap.set(command, annotations)
  } else {
    throw new Error('First help() argument must be a function')
  }
}

export const helper: (
  logger: ILogger,
  argv: string[]
) => Middleware = logger => next => args => {
  const { definition, options, command, namespace } = args
  if (!options.help) {
    return next(args)
  }

  if (command) {
    printCommandHelp(command, namespace, logger)
  }

  if (
    namespace === '' &&
    typeof definition === 'object' &&
    (!command || command === definition.default)
  ) {
    printDefinitionHelp(definition, namespace, logger)
  }
}
