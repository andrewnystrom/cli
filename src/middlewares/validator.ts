import { difference } from 'lodash'

import { ICLIOptions, ICommandFunction } from '../interpreter'
import { optionsToString } from '../utils/options'
import { CLIIllegalOption } from '../utils/errors'

export interface IValidateArguments {
  command: ICommandFunction
  options: ICLIOptions
  namespace: string
}

export function validate({ command, options, namespace }: IValidateArguments) {
  if (typeof command.help !== 'object') {
    return
  }

  const annotations = command.help
  const annotatedOptionsKeys =
    (annotations && annotations.options && Object.keys(annotations.options)) ||
    []
  const optionsKeys = Object.keys(options)
  const illegalOptionsKeys = difference(optionsKeys, annotatedOptionsKeys)

  if (annotatedOptionsKeys.length && illegalOptionsKeys.length) {
    const msg =
      `Illegal option: ${optionsToString(illegalOptionsKeys)}\n` +
      `Available options: ${optionsToString(annotatedOptionsKeys)}\n` +
      `Type "${namespace} --help" for more information`
    throw new CLIIllegalOption(msg)
  }
}
