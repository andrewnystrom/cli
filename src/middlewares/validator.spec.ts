import { annotationsMap, help } from './helper'
import { validator } from './validator'

describe('validator()', () => {
  let args: any
  let next: jest.Mock

  beforeEach(() => {
    annotationsMap.clear()
    next = jest.fn()
    args = {
      command: () => null,
      definition: {
        a: () => null,
        b: () => null
      },
      namespace: '',
      options: {},
      params: []
    }
  })

  describe('when detailed annotations provided', () => {
    beforeEach(() => {
      help(args.command, {
        description: 'General script description',
        options: {
          a: 'description for a option',
          foo: 'description for foo option'
        },
        params: ['abc', 'def']
      })
    })

    it('calls next middleware if provided valid options', () => {
      validator(next)(args)
      expect(next).toHaveBeenCalledWith(args)
      args.options = { a: 1 }
      validator(next)(args)
      expect(next).toHaveBeenCalledWith(args)
    })

    it('throws error if invalid option provided', () => {
      args.options = { abc: 1 }
      expect(() => {
        validator(next)(args)
      }).toThrowErrorMatchingSnapshot()
    })

    it('does not call the next middleware if error thrown', () => {
      args.options = { abc: 1 }
      expect(() => {
        validator(next)(args)
      }).toThrow()
      expect(next).not.toHaveBeenCalled()
    })
  })

  describe('when string as annotation provided', () => {
    beforeEach(() => {
      help(args.command, 'General script description')
    })

    it('calls next middleware', () => {
      validator(next)(args)
      expect(next).toHaveBeenCalledWith(args)
    })
  })

  describe('when annotation not provided', () => {
    it('calls next middleware', () => {
      validator(next)(args)
      expect(next).toHaveBeenCalledWith(args)
    })
  })
})
