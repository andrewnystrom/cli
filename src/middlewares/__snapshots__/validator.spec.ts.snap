// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator() when detailed annotations provided throws error if invalid option provided 1`] = `
"Illegal option: --abc
Available options: -a --foo
Type \\" --help\\" for more information"
`;
